src/iceberg/type.h

```diff
@@ -0,0 +1,397 @@
+class ICEBERG_EXPORT PrimitiveType : public Type {
+ public:
+  bool is_primitive() const override { return true; }
+  bool is_nested() const override { return false; }
+};
+
+/// \brief A data type that may have child fields.
+class ICEBERG_EXPORT NestedType : public Type {
+ public:
+  bool is_primitive() const override { return false; }
+  bool is_nested() const override { return true; }
+
+  /// \brief Get a view of the child fields.
+  [[nodiscard]] virtual std::span<const SchemaField> fields() const = 0;
+  /// \brief Get a field by field ID.
+  [[nodiscard]] virtual std::optional<std::reference_wrapper<const SchemaField>>
+  GetFieldById(int32_t field_id) const = 0;
+  /// \brief Get a field by index.
+  [[nodiscard]] virtual std::optional<std::reference_wrapper<const SchemaField>>
+  GetFieldByIndex(int i) const = 0;
+  /// \brief Get a field by name.
+  [[nodiscard]] virtual std::optional<std::reference_wrapper<const SchemaField>>
+  GetFieldByName(std::string_view name) const = 0;
+};
+
+/// \defgroup type-primitive Primitive Types
+/// Primitive types do not have nested fields.
+/// @{
+
+/// \brief A data type representing a boolean.
+class ICEBERG_EXPORT BooleanType : public PrimitiveType {
+ public:
+  BooleanType() = default;
+  ~BooleanType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 32-bit signed integer.
+class ICEBERG_EXPORT Int32Type : public PrimitiveType {
+ public:
+  Int32Type() = default;
+  ~Int32Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 64-bit signed integer.
+class ICEBERG_EXPORT Int64Type : public PrimitiveType {
+ public:
+  Int64Type() = default;
+  ~Int64Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 32-bit (single precision) float.
+class ICEBERG_EXPORT Float32Type : public PrimitiveType {
+ public:
+  Float32Type() = default;
+  ~Float32Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 64-bit (double precision) float.
+class ICEBERG_EXPORT Float64Type : public PrimitiveType {
+ public:
+  Float64Type() = default;
+  ~Float64Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a fixed-precision decimal.
+class ICEBERG_EXPORT DecimalType : public PrimitiveType {
+ public:
+  constexpr static const int32_t kMaxPrecision = 38;
```

Comment:
Is it worth adding enforcement of constant style to `.clang-tidy`? Now it seems that we prefer `kMaxPrecision` over `MAX_PRECISION`

Reply:
We should add it. Maybe this isn't in line with the style guide

Reply:
This is correct: https://google.github.io/styleguide/cppguide.html#Constant_Names